From: Dan Minor <dminor@mozilla.com>
Date: Wed, 19 Sep 2018 15:06:00 -0400
Subject: Bug 1376873 - Fix up rotation in video_capture_impl.cc; r=pehrsons

This fixes a bug in the upstream code introduced when they removed the
ConvertToI420 helper method from webrtc_libyuv.cc. The buffer size is
passed into libyuv::ConvertI420 incorrectly when rotation is applied, which
causes bad rendering and instabilities.

Differential Revision: https://phabricator.services.mozilla.com/D7478
Mercurial Revision: https://hg.mozilla.org/mozilla-central/rev/c56cb66f86518dfb046c642ebeb07cf51c23a3cb
---
 modules/video_capture/video_capture_impl.cc | 11 ++++++++++-
 1 file changed, 10 insertions(+), 1 deletion(-)

diff --git a/modules/video_capture/video_capture_impl.cc b/modules/video_capture/video_capture_impl.cc
index 4cf8fa366e..82e95892dc 100644
--- a/modules/video_capture/video_capture_impl.cc
+++ b/modules/video_capture/video_capture_impl.cc
@@ -231,10 +231,19 @@ int32_t VideoCaptureImpl::IncomingFrame(uint8_t* videoFrame,
     }
   }
 
+  int dst_width = buffer->width();
+  int dst_height = buffer->height();
+
+  // LibYuv expects pre-rotation_mode values for dst.
+  // Stride values should correspond to the destination values.
+  if (rotation_mode == libyuv::kRotate90 || rotation_mode == libyuv::kRotate270) {
+    std::swap(dst_width, dst_height);
+  }
+
   const int conversionResult = ConvertToI420(
       videoFrame, videoFrameLength, buffer->MutableDataY(), buffer->StrideY(),
       buffer->MutableDataU(), buffer->StrideU(), buffer->MutableDataV(),
-      buffer->StrideV(), width, height, stride_, target_width, target_height,
+      buffer->StrideV(), width, height, stride_, dst_width, dst_height,
       static_cast<uint32_t>(rotation_mode),
       ConvertVideoType(frameInfo.videoType));
   if (conversionResult != 0) {
